pipeline.cpp

```diff
@@ -81,44 +81,43 @@ void DrawLine(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* cons
  ************************************************************/
 void DrawTriangle(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* const attrs, Attributes* const uniforms = NULL, FragmentShader* const frag = NULL)
 {
-    // My Code (with help from the reading)
+    // My Code (with help from the reading and Bro Christiansen)
     // This finds the minimum x and y values of the vectors, creating a bouding box
     // around the triangle that we can loop through
     int maxX = MAX3(triangle[0].x, triangle[1].x, triangle[2].x);
     int minX = MIN3(triangle[0].x, triangle[1].x, triangle[2].x);
     int maxY = MAX3(triangle[0].y, triangle[1].y, triangle[2].y);
     int minY = MIN3(triangle[0].y, triangle[1].y, triangle[2].y);
 
-    // set the origin of the colliding vectors to 0, 0. This makes the cross-product work much more simply
+    // set the origin of each vector to 0. This makes the cross-product/determinant much easier
     Vertex * vert1 = new Vertex {triangle[1].x - triangle[0].x, triangle[1].y - triangle[0].y, 1, 1};
-    Vertex * vert2 = new Vertex {triangle[2].x - triangle[0].x, triangle[2].y - triangle[0].y, 1, 1};
+    Vertex * vert2 = new Vertex {triangle[2].x - triangle[1].x, triangle[2].y - triangle[1].y, 1, 1};
+    Vertex * vert3 = new Vertex {triangle[0].x - triangle[2].x, triangle[0].y - triangle[2].y, 1, 1};
+
+    // find the total area of the triangle, we need this to find relative areas later
+    double totArea = Determinant(vert1->x, vert1->y, vert2->x, vert2->y);
 
     // looping through each pixel in the bounding box
     for (int x = minX; x <= maxX; x++)
     {
         for (int y = minY; y <= maxY; y++)
         {
-            // create a temporary vertex at current pixel in the square
-            Vertex * q = new Vertex {x - triangle[0].x, y - triangle[0].y, 1, 1};
```

Comment:
I did this way too, and it isn't the most efficient. So using Bro Christensen's algorithm was a good idea.